design-proposals/003-credentials.md

```diff
@@ -0,0 +1,226 @@
+
+NOTE: Salesforce supports using x509 certificates for OAuth flow, and that should be preferred as the refreshing OAuth token can be performed within certificate's expiry i.e. not depending on the refresh token.
+
+### Credential view model
+
+Each credential should have a name so that the user can identify it, type: this can be username-password, x509 or OAuth tokens, internal identifier and connector label so it can be determined to what connection it can be applied to.
+
+The credential name can be auto generated (e.g. "Salesforce credentials for user@example.com").
+
+View model credential properties:
+
+ * **id**        : System assigned identifier for the credential, e.g. UUID
+ * **name**      : Name of the credential, for humans, e.g. "Salesforce credentials for user@example.com"
+ * **connector** : Connector type (label) that the credential applies to, e.g. "salesforce"
+ * **created**   : Date/time the credential was created
+ * **note**      : Optional note, could be generated during credential acquisition, e.g. "Expires 23.11.2018"
+ 
+This constitutes the view model, notice that the sensitive (secret) data is not exposed to the UI -- the UI should reference credentials by the identifier.
+
+The view model is generated by the backend from kubernetes secrets, mapping secret name to connection, attributes above as annotations of the secret.
+
+### Credential backend model
+
+As credential providers would share common implementation details, surely for application of credentials, i.e. creating the kubernetes secret from acquired data, this facilitates the need for a common credential backend model. This backend model can be viewed as data transfer object between credential provider, logic that handles the application of credentials and persistence in kubernetes secrets.
+
+The backend model is essentially a set of key-value pairs identified by the same **id** to be used in the view layer. The keys correspond to property names needed in connector configuration (i.e. set in `application.properties`) and the values are obtained during credential acquisition.
+
+As some secrets are not exposed to the UI, e.g. secrets that are private to Syndesis such as OAuth client secret or secret/private key, the same backend model constructed during the creation of kubernetes deployment descriptor (i.e. project generation).
+
+
+## Implementation
+
+The Credential provider should implement the following Java interface:
+
+```java
+interface CredentialProvider {
+    /**
+     * All AcquisitionMethod that this CredentialProvider supports
+     */
+    AcquisitionMethod[] acquisitionMethods();
+    /**
+     * Begin acquisition of the credential by returning the information
+     * to the UI on how to proceed and make any arrangements needed 
+     */
+    Acquisition acquire(String selectedMethod, Map<String, String> methodParameters, ContinuationStore continuationStore);
```

Comment:
@rhuss basically some form of session, we need to persist `redirectUrl` given by the UI, it could be done stateless if the SaaS has an option to pass along value via request parameters to CredentialProvider, but that could lead to tampering or open redirect security issues.